WIP: Stage1

KB00100100 opened the pull request:
Streamr-client-python stage1.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/streamr-dev/streamr-client-python?pullRequest=3) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/streamr-dev/streamr-client-python?pullRequest=3) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/streamr-dev/streamr-client-python?pullRequest=3) it.</sub>

githubwei commented on streamr/client/client.py:
```diff
@@ -0,0 +1,398 @@
+        if apiKey == None:
+            apiKey = self.options['auth']['apiKey']
+            
+        streamId = objectOrId.streamId
```
I would suggest adding option to publish by providing simple streamId string, in addition to getting it from object attribute.
```
if hasattr(objectOrId, 'streamId'):
    streamId = objectOrId.streamId
else:
    streamId = objectOrId
```

KB00100100 replied:
Thanks for your suggestion, I will revise it later.

mikhaelsantos commented:
streamr/protocol/payloads.py:1:59: E203 whitespace before ','
streamr/protocol/payloads.py:26:31: E711 comparison to None should be 'if cond is not None:'
streamr/protocol/payloads.py:34:22: E271 multiple spaces after keyword
streamr/protocol/payloads.py:38:1: W293 blank line contains whitespace
streamr/protocol/payloads.py:47:61: E225 missing whitespace around operator
streamr/protocol/payloads.py:49:47: E251 unexpected spaces around keyword / parameter equals
streamr/protocol/payloads.py:60:21: E128 continuation line under-indented for visual indent
streamr/protocol/payloads.py:67:21: E124 closing bracket does not match visual indentation
streamr/protocol/payloads.py:109:13: E122 continuation line missing indentation or outdented
streamr/protocol/request.py:27:89: E202 whitespace before ')'
streamr/protocol/request.py:147:5: E301 expected 1 blank line, found 0
streamr/protocol/utils/parser.py:24:9: E722 do not use bare 'except'
streamr/client/client.py:96:29: E221 multiple spaces before operator
streamr/client/client.py:122:38: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
streamr/client/event.py:21:69: E227 missing whitespace around bitwise or shift operator
streamr/client/subscription.py:96:50: E272 multiple spaces before keyword
streamr/client/subscription.py:102:89: E201 whitespace after '('
streamr/client/util/websock.py:172:37: E127 continuation line over-indented for visual indent
tests/unit/connection_test.py:58:26: W292 no newline at end of file
tests/unit/response_test.py:92:30: E222 multiple spaces after operator
tests/unit/subscription_test.py:13:26: E721 do not compare types, use 'isinstance()'

kare commented:
@mikhaelsantos What tool produces the output above?

mikhaelsantos commented:
requirements.txt file with dependencies would be nice.

KB00100100 commented:
> Python style guide checker
> Tool: pycodestyle
> 
> streamr/protocol/payloads.py:1:59: E203 whitespace before ','
> streamr/protocol/payloads.py:26:31: E711 comparison to None should be 'if cond is not None:'
> streamr/protocol/payloads.py:34:22: E271 multiple spaces after keyword
> streamr/protocol/payloads.py:38:1: W293 blank line contains whitespace
> streamr/protocol/payloads.py:47:61: E225 missing whitespace around operator
> streamr/protocol/payloads.py:49:47: E251 unexpected spaces around keyword / parameter equals
> streamr/protocol/payloads.py:60:21: E128 continuation line under-indented for visual indent
> streamr/protocol/payloads.py:67:21: E124 closing bracket does not match visual indentation
> streamr/protocol/payloads.py:109:13: E122 continuation line missing indentation or outdented
> streamr/protocol/request.py:27:89: E202 whitespace before ')'
> streamr/protocol/request.py:147:5: E301 expected 1 blank line, found 0
> streamr/protocol/utils/parser.py:24:9: E722 do not use bare 'except'
> streamr/client/client.py:96:29: E221 multiple spaces before operator
> streamr/client/client.py:122:38: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
> streamr/client/event.py:21:69: E227 missing whitespace around bitwise or shift operator
> streamr/client/subscription.py:96:50: E272 multiple spaces before keyword
> streamr/client/subscription.py:102:89: E201 whitespace after '('
> streamr/client/util/websock.py:172:37: E127 continuation line over-indented for visual indent
> tests/unit/connection_test.py:58:26: W292 no newline at end of file
> tests/unit/response_test.py:92:30: E222 multiple spaces after operator
> tests/unit/subscription_test.py:13:26: E721 do not compare types, use 'isinstance()'

Thanks for your kindly suggestions, I will revise them later.

KB00100100 commented:
Hello, Streamr friends. We revised the python library, and the following shows the summary of the revision.

# githubwei
```
#Question1: 
I would suggest adding option to publish by providing simple streamId string, in addition to getting it from object attribute.

if hasattr(objectOrId, 'streamId'):
    streamId = objectOrId.streamId
else:
    streamId = objectOrId

#Answer1:
added this option.
```

# mikhaelsantos
```
#Question1: 
Python style guide checker
Tool: pycodestyle

streamr/protocol/payloads.py:1:59: E203 whitespace before ','
streamr/protocol/payloads.py:26:31: E711 comparison to None should be 'if cond is not None:'
streamr/protocol/payloads.py:34:22: E271 multiple spaces after keyword
streamr/protocol/payloads.py:38:1: W293 blank line contains whitespace
streamr/protocol/payloads.py:47:61: E225 missing whitespace around operator
streamr/protocol/payloads.py:49:47: E251 unexpected spaces around keyword / parameter equals
streamr/protocol/payloads.py:60:21: E128 continuation line under-indented for visual indent
streamr/protocol/payloads.py:67:21: E124 closing bracket does not match visual indentation
streamr/protocol/payloads.py:109:13: E122 continuation line missing indentation or outdented
streamr/protocol/request.py:27:89: E202 whitespace before ')'
streamr/protocol/request.py:147:5: E301 expected 1 blank line, found 0
streamr/protocol/utils/parser.py:24:9: E722 do not use bare 'except'
streamr/client/client.py:96:29: E221 multiple spaces before operator
streamr/client/client.py:122:38: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
streamr/client/event.py:21:69: E227 missing whitespace around bitwise or shift operator
streamr/client/subscription.py:96:50: E272 multiple spaces before keyword
streamr/client/subscription.py:102:89: E201 whitespace after '('
streamr/client/util/websock.py:172:37: E127 continuation line over-indented for visual indent
tests/unit/connection_test.py:58:26: W292 no newline at end of file
tests/unit/response_test.py:92:30: E222 multiple spaces after operator
tests/unit/subscription_test.py:13:26: E721 do not compare types, use 'isinstance()'

#Answer1:
revised.

#Question2: 
streamr/client/subscription.py:3:0: W0614: Unused import UnsupportedVersionError from wildcard import (unused-wildcard-import) streamr/client/connection.py:10:0: W0611: Unused import time (unused-import) streamr/client/connection.py:11:0: W0611: Unused import os (unused-import) streamr/client/connection.py:12:0: W0611: Unused import json (unused-import) streamr/client/connection.py:90:25: W0613: Unused argument 'ws' (unused-argument)

Also attributes don't confirm with snake case and constants don't confirm with upper case naming.

pylint rating -0.53/10

#Answer2:
Thanks for your kindly suggentions. We are not clear this problem must be revised? We perfer that all the functions and attributes are in camel case.
```

# mirpo
```
#Question1:
Consider using pure pytest instead of Jupiter

#Answer1:
We added pytest examples: create_client.py connect_and_publish.py create_or_get_stream.py and publish remove jupyterfilenotebook

#Question2:
More examples and not just copied tests. Separate examples how to create, subscribe and publish

#Answer2:
We added pytest examples: create_client.py connect_and_publish.py create_or_get_stream.py and publish remove jupyterfilenotebook

#Question3:
Move from code and keep all API keys to .env files

#Answer3:
Revised.

#Question4:
+from streamr.client.event import Event
+from streamr.client.connection import Connection
+from streamr.rest.session import getSeToken_APIKey
Function name should be in lowercase

#Answer4:
Event and Connection are classes and  we changed the function getSeToken_APIKey to the camel style.

#Question5:
Remove all unused imports

#Answer5:
We removed all ununsed imports.

#Question6:
self.options = {**self.options,**options}
if self.options.get('apiKey',None) == None and self.options.get('authKey',None) == None and self.options.get('auth',{}).get('apiKey',None) == None:

Consider using some Options class to work with options instead of working with an array

#Answer6:
We will fix this at the second stage. 

#Question7:
self.sessionThread = None
self.__autoUpdateSessionToken()
if self.sessionToken  == None:

Correct way is:
if self.sessionToken is None
extends to other conditions

#Answer7:
Revised.

#Question8:
except:
     pass
else:
     raise Exception('Wait for conneted event before calling publish or set autoConnect to True')

Typo and using just Exception is not correct. You should different types https://docs.python.org/3/library/exceptions.html
For example for websocket you should use exceptions defined in library websocket/_exceptions.py

#Answer8:
We defined some Exception classes and the code will raise those Exceptions.

#Question9:
from setuptools import setup

It's better to create requirements.txt as well (https://caremad.io/posts/2013/07/setup-vs-requirement/)

#Answer9:
The requirement.txt is in setup.py and we also added the requirement.txt.

#Question10:
if callback in self.event_list[event_name]:
    self.event_list[event_name].remove(callback)
else:
    logging.warning('Event: %s has no callback function: %s'&(event_name,callback))

& - error

#Answer10:
Revised.

#Question11:
import logging
class Event():

class Event:

#Answer11:
Revised.

#Question12:
cli = Client({'autoConnect': False, 'autoDisconnect': False}, conn)

error - no ApiKey

#Answer12:
Revised.

#Question13:
optlist, args = getopt.getopt(sys.argv[1:], 'i:j:k:')
for m,t in optlist:
    if m == '-i':

Hard to understand logic, name variables according actions

#Answer13:
Revised. We split the code into many small functions.

#Question14:
Consider splitting test into small functions.
Example:

class TestStreamMessage:
    def stream_message_v28:
        # given
        arr = [28, 'TsvTbqshTsuLg_HyUjxigA', 0, 1529549961116, 0, 941516902, 941499898, StreamMessage.CONTENT_TYPE.JSON, '{"valid": "json"}']

        # when
        msg = StreamMessage.deserialize(arr)

        # then
        assert isinstance(msg,StreamMessage)
        ...etc
    
    def stream_message_v29:    
    

etc

and move helper functions to the top of each test file

#Answer14:
Revised.

#Question15:
Additional notes:

Missing gitignore
Using free PyCharm will help to reformat and follow Python code standards
Configure Travis CI run tests

#Answer15:
* We added the gitignore file.
* Yes, we do use tools to reformat the codes.
* Because we don't have experience about the Travis CI, we need more time to use this tools and we will finish this when stage 2 is completed.
```

mirpo commented on examples/create_or_get_Stream.py:
```diff
@@ -0,0 +1,23 @@
+from streamr.client.client import Client
```
File name create_or_get_Stream.py....

KB00100100 replied:
We revised the file name.

mirpo commented on examples/create_or_get_Stream.py:
```diff
@@ -0,0 +1,23 @@
+
+# get streamById will return the stream with the streamId
+# the return is a dictionary containning the information of the stream
+myClient.getStreamById('streamId')
```
Test created new stream by failed to get "Fail to Get a Stream using its ID. The Status code: 404"

KB00100100 replied:
We added a comment for the method. The error happens due to the wrong Stream ID. You should replace the streamId before using the method. The streamId can be found in the stream page, also can be obtained using the 'getStreamByName' method of which the response is shown as the following:
```
[{'config': {'fields': []},
  'dateCreated': '2019-01-27T09:16:22Z',
  'description': 'None',
  'feed': {'id': 7, 'module': 147, 'name': 'API'},
  'id': 'NQUihKiZTT6cc9F2-bLU2A',
  'lastUpdated': '2019-01-27T09:16:22Z',
  'name': 'stream-test',
  'partitions': 1,
  'requireSignedData': False,
  'uiChannel': False}]
```

mirpo commented on examples/create_or_get_Stream.py:
```diff
@@ -0,0 +1,23 @@
+# Note that createStream is running forcefully.
+# That means you can create two stream with same name but the streamIds are different
+
+
```
add stream_id = 'stream-test'

KB00100100 replied:
We are not sure what do you mean. We think this issue is the same as the previous one. The streamId should be a 22 bytes string, which can be found in the stream page, also can be obtained using the 'getStreamByName' method.

mirpo commented on examples/connect_and_publish.py:
```diff
@@ -0,0 +1,34 @@
+
+# To connect to the server, you can use connect method
+# This method returns immediately  and you can query the state by isConnected funciton
+myClient.connect()
```
Failed to connect - "raise Exception('Already connecting')"

KB00100100 replied:
This happens due to the autoconnect attribute of the client. We disabled the autoconnect and autodisconnect options.
```
myOption = {'apiKey': 'your-api-key',
            'autoConnect': False, 'autoDisconnect': False}
```

mirpo commented on streamr/client/client.py:
```diff
@@ -0,0 +1,416 @@
+
+    def getOrCreateStream(self, streamname):
+        stream = self.getStreamByName(streamname)
+        if len(stream) == 0 or stream is None:
```
Only "stream is None" is needed, len(stream) will rise an error "TypeError: object of type 'NoneType' has no len()"

KB00100100 replied:
We used len(stream) because a list type without any element is returned when using the getStreamByName method, rather than a None type. We revised the codes as the following.
```
    def getOrCreateStream(self, streamname):
        stream = self.getStreamByName(streamname)
        if stream is None or (isinstance(stream,list) and len(stream) == 0 ):
            return self.createStream(streamname)
        else:
            return stream
```

mirpo commented:
Report from pylint 
[pylint.txt](https://github.com/streamr-dev/streamr-client-python/files/2936359/pylint.txt)

KB00100100 commented:
Hello, wei and mirr. We revised the python library. In this revision, we modified the library to follow PEP8 standard, also fixed some bugs. The following picture shows the inspection results from the Pycharm IDE.
![image](https://user-images.githubusercontent.com/37429381/54273166-45499b80-45c0-11e9-8bdd-a4f2ec34c501.png)
